File: ActivityLauncherApp/src/main/java/de/szalkowski/activitylauncher/async/IconListAsyncProvider.java
package de.szalkowski.activitylauncher.async;

import android.content.Context;

import de.szalkowski.activitylauncher.ui.adapter.IconListAdapter;

public class IconListAsyncProvider extends AsyncProvider<IconListAdapter> {
    private final IconListAdapter adapter;

    public IconListAsyncProvider(Context context, Listener<IconListAdapter> listener) {
        super(context, listener, false);
        this.adapter = new IconListAdapter(context);
    }

    @Override
    protected IconListAdapter run(Updater updater) {
        adapter.resolve(updater);
        return this.adapter;
    }
}

File: app/src/main/java/de/szalkowski/activitylauncher/ui/IconListAsyncProvider.kt
package de.szalkowski.activitylauncher.ui

import android.content.Context
import dagger.assisted.Assisted
import dagger.assisted.AssistedFactory
import dagger.assisted.AssistedInject
import dagger.hilt.android.qualifiers.ActivityContext

class IconListAsyncProvider @AssistedInject constructor(
    @ActivityContext context: Context,
    private val adapter: IconListAdapter,
    @Assisted listener: Listener<IconListAdapter>?,
) : AsyncProvider<IconListAdapter>(context, listener, false) {
    @AssistedFactory
    interface IconListAsyncProviderFactory {
        fun create(listener: Listener<IconListAdapter>?): IconListAsyncProvider
    }

    override fun run(updater: Updater?): IconListAdapter {
        adapter.resolve(updater)
        return this.adapter
    }
}
